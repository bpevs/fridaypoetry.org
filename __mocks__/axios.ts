const get = jest.fn((url: string) => {
  return Promise.resolve({ data: {
    poem1: "poem1",
    poem2: "poem2",
    poem3: "poem3",
  }});
});

const post = jest.fn((url: string, poemData: any) => {
  if (!poemData) {
    return Promise.reject("No poem Data");
  }
  const { author, text, title } = poemData;

  if (typeof text !== "string") {
    return Promise.reject("no poem text");
  }

  return Promise.resolve({data: { name: "poem-id" }});
});

export default { get, post };
