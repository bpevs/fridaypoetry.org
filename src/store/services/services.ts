const poems = require("./__mocks__/fixtures/poems.json");
const baseUrl = "https://friday-poetry.firebaseio.com/";

export function createPoem(poem: any) {
  return Promise.resolve();
}

export function fetchPoem(poemId: number) {
  return fetch(`${baseUrl}/poems/${poemId}.json`);
}

export function fetchPoems(startId?: number) {
  const url = startId == null
    ? `${baseUrl}/poems.json?orderBy="id"`
    : `${baseUrl}/poems.json?orderBy="id"&startAt=${startId - 5}&endAt=${startId + 5}`;

  return fetch(url)
    .then(response => response.json() )
    .then(poems => ({ poems }));
}
