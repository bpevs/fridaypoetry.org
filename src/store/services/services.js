import axios from "axios";
import { pick } from "lodash";
import { BASE_URL } from "../../config";


export function createPoem(poem) {
  const IS_FRIDAY = (new Date()).getDay() === 5;

  if (!IS_FRIDAY) {
    alert("Oops! We can't submit this poem! It's not Friday!");
    return Promise.reject("It's not Friday!");
  }

  if (!poem.text || (typeof poem.text !== "string")) {
    return Promise.reject("No Poem text!");
  }

  const poemData = pick(poem, [ "author", "text", "title" ]);
  return axios.post(`${BASE_URL}.json`, poemData)
    .then((response) => ({ poemId: response.data.name }));
}

export function fetchPoemIds() {
  return axios.get(`${BASE_URL}.json?shallow=true`)
    .then((response) => {
      const ids = Object.keys(response.data).sort().reverse();
      return { ids };
    });
}

export function fetchPoem(poemId) {
  return axios.get(`${BASE_URL}/${poemId}.json`)
    .then((response) => ({ poem: response.data }));
}
