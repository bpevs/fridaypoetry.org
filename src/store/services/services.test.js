jest.mock("axios");

import axios from "axios";
import { BASE_URL } from "../../config";
import { createPoem, fetchPoem, fetchPoemIds } from "./services";


afterEach(() => {
  jest.clearAllMocks();
});

describe("createPoem", function() {
  beforeEach(() => {
    Date.prototype.getDay = () => 5;

    this.poem = {
      author: "author",
      text: "text",
      title: "title"
    };
  });

  test("Should create a poem if it's friday", () => {
    return createPoem(this.poem)
      .then(({ poemId }) => {
        expect(typeof poemId).toEqual("string");
        expect(axios.post.mock.calls.length).toEqual(1);
        const [ url, poemData ] = axios.post.mock.calls[0];
        expect(url).toEqual(`${BASE_URL}.json`);
        expect(poemData).toMatchSnapshot();
      });
  });

  test("Should create a poem without an author or title if it's friday", () => {
    this.poem.author = undefined;
    this.poem.title = undefined;

    return createPoem(this.poem)
      .then(({ poemId }) => {
        expect(typeof poemId).toEqual("string");
        expect(axios.post.mock.calls.length).toEqual(1);
        const [ url, poemData ] = axios.post.mock.calls[0];
        expect(url).toEqual(`${BASE_URL}.json`);
        expect(poemData).toMatchSnapshot();
      });
  });

  test("Should fail on friday if no text", () => {
    this.poem.text = undefined;

    return createPoem(this.poem)
      .catch(error => {
        expect(error).toEqual("No Poem text!");
        expect(axios.post.mock.calls.length).toEqual(0);
      });
  });

  test("Should fail if it's not friday", () => {
    Date.prototype.getDay = () => 4;

    return createPoem(this.poem)
      .catch(error => {
        expect(error).toEqual("It's not Friday!");
        expect(axios.post.mock.calls.length).toEqual(0);
      });
  });
});

describe("fetchPoemIds", () => {
  test("Should fetch poem ids", () => {
    return fetchPoemIds()
      .then(({ ids }) => {
        expect(axios.get.mock.calls.length).toEqual(1);
        const [ url ] = axios.get.mock.calls[0];
        expect(url).toEqual(`${BASE_URL}.json?shallow=true`);
        expect(Array.isArray(ids)).toEqual(true);
      });
  });
});

describe("fetchPoem", () => {
  test("Should fetch poem", () => {
    const poemId = "success-poem";
    return fetchPoem(poemId)
      .then(poem => {
        expect(axios.get.mock.calls.length).toEqual(1);
        const [ url ] = axios.get.mock.calls[0];
        expect(url).toEqual(`${BASE_URL}/${poemId}.json`);
      });
  });
});
