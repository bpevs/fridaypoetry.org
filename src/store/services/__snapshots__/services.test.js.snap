// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPoem Should create a poem if it's friday 1`] = `
Object {
  "author": "author",
  "text": "text",
  "title": "title",
}
`;

exports[`createPoem Should create a poem without an author or title if it's friday 1`] = `
Object {
  "author": undefined,
  "text": "text",
  "title": undefined,
}
`;
